### PR TITLE
In-memory caching for event readers

### LINKED ISSUE #2: Middleware page caching

### PR DESCRIPTION
**Pull request by Grepsy**
Initial draft for an in-memory caching layer that wraps an IEventFeedReader based on https://github.com/dburriss/event-feed/issues/2.

Some considerations:

- Complete pages now still expire within an absolute expiration time. It would be nicer if we could allocate a fixed size memory budget to keep pages in memory as long as we're within budget but this seems tricky as calculating object memory usage is required to set the `Size` property on `MemoryCacheEntryOptions`. A possible solution would be to estimate this based on the payload size.
- Usage-metrics are kept out of scope but this impl. should not prevent that.
- No worker-based cache refreshing is done, see https://github.com/dburriss/event-feed/issues/2#issuecomment-1092052140.
- F# might not be idiomatic, feedback appreciated!

Further work:

- Tests + docs
- ASP.NET Core helpers to easily add this.
- Make expiration options configurable.
- Support new event notification and update the in-memory cache.
- Possibly abstract the IMemoryCache to support distributed caching scenarios (seems like this can be done when actually required).
- Discuss back-channel notifications (needs more information).

### CODE REVIEW

File: src/EventFeed.Producer.Caching/CachingEventFeedReader.fs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    let getCacheEntryOptions (pageNumber : int) =
+        match isLastPage(pageNumber) with
+        | true ->  new MemoryCacheEntryOptions(AbsoluteExpirationRelativeToNow = lastPageExpirationTime)
+        // Would be nice if we can use Size here so completed pages only get evicted when cache reaches max size
+        | false -> new MemoryCacheEntryOptions(AbsoluteExpirationRelativeToNow = completePageExpirationTime )
+
+    interface IEventFeedReader with
+
+        member this.EventNumbers() = underlyingReader.EventNumbers()
```
Comment (dburriss) [inline src/EventFeed.Producer.Caching/CachingEventFeedReader.fs:26]: Small one but maybe `innerReader`?

File: src/EventFeed.Producer.Caching/CachingEventFeedReader.fs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+        member this.ReadPage pageNumber =
+            match cache.Get(pageNumber) with
+            | :? seq<FeedEvent> as entry -> entry
+            | null ->
```
Comment (dburriss) [inline src/EventFeed.Producer.Caching/CachingEventFeedReader.fs:31]: `match` is checked in order, so check `null` first.

File: src/EventFeed.Producer.Caching/CachingEventFeedReader.fs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        | true ->  new MemoryCacheEntryOptions(AbsoluteExpirationRelativeToNow = lastPageExpirationTime)
+        // Would be nice if we can use Size here so completed pages only get evicted when cache reaches max size
+        | false -> new MemoryCacheEntryOptions(AbsoluteExpirationRelativeToNow = completePageExpirationTime )
+
+    interface IEventFeedReader with
+
+        member this.EventNumbers() = underlyingReader.EventNumbers()
+
+        member this.ReadPage pageNumber =
+            match cache.Get(pageNumber) with
```
Comment (dburriss) [inline src/EventFeed.Producer.Caching/CachingEventFeedReader.fs:29]: `Get` has a `Get<T>` overload right? Could use that and not have to do the type check?
What may be more idiomatic but have a slight perf hit so wouldn't worry about it is having a closure over `cache` that returns an `Option` and then match on `Some` and `None`.
Or we can do it and tune later when we have benchmarks :)

**Review by dburriss — CHANGES_REQUESTED** [review]

File: src/EventFeed.Producer.Caching/CachingEventFeedReader.fs (hunk 3 above)
Comment (Grepsy) [reply src/EventFeed.Producer.Caching/CachingEventFeedReader.fs:29]: Yes, you're right! I was looking for that but thought I was mistaken, but for some reason these extension methods are not showing up in my IntelliSense in F#. I also don't get auto import, or rename refactoring support. Is that normal or my setup?

File: src/EventFeed.AspNetCore/EventFeedServicesExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            var builder = new EventFeedOptionsBuilder();
+            var options = builder.Options;
+            optionsBuilder.Invoke(builder);
+
+            if (options.CacheEnabled)
+            {
+                // wrap the EventFeedReader in a caching layer
+                serviceCollection.AddTransient<IEventFeedReader>((provider) => new CachingEventFeedReader(
+                    options.EventFeedReader,
+                    provider.GetRequiredService<IMemoryCache>(),
```
Comment (Grepsy) [inline src/EventFeed.AspNetCore/EventFeedServicesExtensions.cs:21]: Document this is required.

File: test/building/EventFeed.Producer.Caching.Tests.Unit/EventFeed.Producer.Caching.Tests.Unit.fsproj [hunk 5]
```diff
@@ -0,0 +1,38 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net6.0</TargetFramework>
+
```
Comment (Grepsy) [inline test/building/EventFeed.Producer.Caching.Tests.Unit/EventFeed.Producer.Caching.Tests.Unit.fsproj:5]: Remove line.

**Review by dburriss — APPROVED** [review]
We went through together. You made notes.

**Review by dburriss — APPROVED** [review]

**Pull request closed without merging**